### PR TITLE
[IOTDB-2807]Speed up the cross space compaction by multi-threads

### PR DESCRIPTION
**Pull request by choubenson**
The cross space comapction execution performance of 0.13.0 is slower than 0.12.5.

Need to speed up.

We achieve parallel reading and writing at the measurement level by adding threads.

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/CompactionUtils.java [hunk 1]
```diff
@@ -268,6 +274,27 @@ private static AbstractCompactionWriter getCompactionWriter(
     }
   }
 
+  private static void updateDeviceStartTimeAndEndTime(
+      List<TsFileResource> targetResources, AbstractCompactionWriter compactionWriter) {
+    List<TsFileIOWriter> targetFileWriters = compactionWriter.getFileIOWriter();
+    for (int i = 0; i < targetFileWriters.size(); i++) {
+      TsFileIOWriter fileIOWriter = targetFileWriters.get(i);
+      TsFileResource fileResource = targetResources.get(i);
+      // skip the target file that has been deleted
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/CompactionUtils.java:283]: why deleted, add comment

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/writer/CrossSpaceCompactionWriter.java [hunk 2]
```diff
@@ -36,36 +36,34 @@
   // source tsfiles
   private List<TsFileResource> seqTsFileResources;
 
-  private int seqFileIndex;
+  // subTaskId -> seqFileIndex
+  private Map<Integer, Integer> seqFileIndexMap;
```
[the remark this answers is not in the document]
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/writer/CrossSpaceCompactionWriter.java:40]: Resolved

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/writer/AbstractCompactionWriter.java [hunk 3]
```diff
@@ -50,34 +52,34 @@
       MetricConfigDescriptor.getInstance().getMetricConfig().getEnableMetric();
 
   // point count in current measurment, which is used to check size
-  private int measurementPointCount;
+  private Map<Integer, Integer> measurementPointCountMap = new ConcurrentHashMap<>();
```
[the remark this answers is not in the document]
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/writer/AbstractCompactionWriter.java:55]: Resolved

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/writer/AbstractCompactionWriter.java [hunk 4]
```diff
@@ -50,34 +52,34 @@
       MetricConfigDescriptor.getInstance().getMetricConfig().getEnableMetric();
 
   // point count in current measurment, which is used to check size
```
[the remark this answers is not in the document]
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/writer/AbstractCompactionWriter.java:54]: Resolved

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/writer/CrossSpaceCompactionWriter.java [hunk 5]
```diff
@@ -83,21 +81,25 @@ public void endChunkGroup() throws IOException {
...
-    chunkWriter = null;
-    seqFileIndex = 0;
+  public void endMeasurement(int subTaskId) throws IOException {
+    writeRateLimit(chunkWriterMap.get(subTaskId).estimateMaxSeriesMemSize());
+    synchronized (fileWriterList.get(seqFileIndexMap.get(subTaskId))) {
+      chunkWriterMap
+          .get(subTaskId)
+          .writeToFileWriter(fileWriterList.get(seqFileIndexMap.get(subTaskId)));
+    }
+    // chunkWriterMap.get(subTaskId)=null;
```
[the remark this answers is not in the document]
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/writer/CrossSpaceCompactionWriter.java:91]: Resolved

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/writer/InnerSpaceCompactionWriter.java [hunk 6]
```diff
@@ -80,7 +79,13 @@ public void close() throws IOException {
     if (fileWriter != null && fileWriter.canWrite()) {
       fileWriter.close();
     }
-    chunkWriter = null;
+    // chunkWriter = null;
```
[the remark this answers is not in the document]
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/writer/InnerSpaceCompactionWriter.java:82]: Resolved

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/writer/InnerSpaceCompactionWriter.java [hunk 7]
```diff
@@ -50,17 +48,18 @@ public void endChunkGroup() throws IOException {
...
-  public void endMeasurement() throws IOException {
-    writeRateLimit(chunkWriter.estimateMaxSeriesMemSize());
-    chunkWriter.writeToFileWriter(fileWriter);
-    chunkWriter = null;
+  public void endMeasurement(int subTaskId) throws IOException {
+    writeRateLimit(chunkWriterMap.get(subTaskId).estimateMaxSeriesMemSize());
+    synchronized (fileWriter) {
+      chunkWriterMap.get(subTaskId).writeToFileWriter(fileWriter);
+    }
+    // chunkWriter = null;
```
[the remark this answers is not in the document]
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/writer/InnerSpaceCompactionWriter.java:56]: Resolved

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/CompactionUtils.java (hunk 1 above)
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/CompactionUtils.java]: Resolved

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/writer/AbstractCompactionWriter.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/writer/AbstractCompactionWriter.java]: Resolved

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/CompactionUtils.java [hunk 8]
```diff
@@ -170,59 +178,58 @@ private static void compactNonAlignedSeries(
...
-        if (IoTDBDescriptor.getInstance().getConfig().isEnableIDTable()) {
-          measurementSchemas.add(IDTableManager.getInstance().getSeriesSchema(device, measurement));
-        } else {
-          measurementSchemas.add(
-              IoTDB.schemaProcessor.getSeriesSchema(new PartialPath(device, measurement)));
-        }
-      } catch (PathNotExistException e) {
-        logger.info("A deleted path is skipped: {}", e.getMessage());
-        continue;
+      if (measurementsForEachSubTask[idx % subTaskNums] == null) {
```
Comment (THUMarkLau) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/CompactionUtils.java:191]: Is it necessary to hash here? If the hash is not uniform, then the effect of multithreading could not be worse. Why not just use a concurrency stack or queue, each sub-thread get a measurement to compact when it is available.

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/CompactionUtils.java [hunk 9]
```diff
@@ -170,59 +178,58 @@ private static void compactNonAlignedSeries(
...
-        writeWithReader(compactionWriter, dataBatchReader);
-        compactionWriter.endMeasurement();
+    // wait for all sub tasks finish
+    for (int i = 0; i < subTaskNums; i++) {
+      try {
+        futures.get(i).get();
+      } catch (InterruptedException | ExecutionException e) {
+        logger.error("SubCompactionTask meet errors ", e);
+        Thread.interrupted();
+        throw new IOException(e);
```
Comment (THUMarkLau) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/CompactionUtils.java:220]: Throw a InterruptedException instead.

**Review by THUMarkLau — CHANGES_REQUESTED** [review]
Do we need to have a performance report for this PR before it is merged?

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/CompactionUtils.java (hunk 9 above)
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/CompactionUtils.java:220]: Resolved

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/CompactionUtils.java (hunk 8 above)
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/CompactionUtils.java:191]: Each sub task has its own hashset, so this won't happen.

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/writer/CrossSpaceCompactionWriter.java [hunk 10]
```diff
@@ -169,10 +181,34 @@ private void checkIsDeviceExistAndGetDeviceEndTime() throws IOException {
...
+      fileWriterList.get(fileIndex).startChunkGroup(deviceId);
+      hasTargetFileStartChunkGroup[fileIndex].set(2);
+    }
+  }
+
+  private void flushChunkToFileWriter(int subTaskId) throws IOException {
+    int fileIndex = seqFileIndexArray[subTaskId];
+    writeRateLimit(chunkWriterMap.get(subTaskId).estimateMaxSeriesMemSize());
+    while (hasTargetFileStartChunkGroup[fileIndex].get() == 1) {
+      // wait until the target file has finished flushing chunk group header
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/writer/CrossSpaceCompactionWriter.java:196]: wait  and notifyall

File: tsfile/src/main/java/org/apache/iotdb/tsfile/write/writer/TsFileIOWriter.java [hunk 11]
```diff
@@ -458,6 +458,10 @@ public void writePlanIndices() throws IOException {
     out.flush();
   }
 
+  public void truncate(long truncateSize) throws IOException {
```
Comment (qiaojialin) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/write/writer/TsFileIOWriter.java:461]:
```suggestion
  public void truncate(long offset) throws IOException {
```

Comment (choubenson) [reply tsfile/src/main/java/org/apache/iotdb/tsfile/write/writer/TsFileIOWriter.java]: Resolved

**Review by qiaojialin — APPROVED** [review]

**Pull request merged**